Update coupons to be specific

zingsurfer opened the pull request:
Before, coupons have been applied to any offer item generically. Since offered items are starting to vary greatly, we'd like to specify which certifications(s) a coupon is for. 

This update adapts the couponsgem to generate coupons that are specific to a subset of items, instead of coupons that apply to any item. Generic coupons will still work until they are expired or updated to be specific.

I kept this minimal to start. We could also update the generated views here too.

zingsurfer committed:
Add polymorphic offerable association

Allows apps that use this gem can apply coupons to different model
objects.

zingsurfer committed:
Update coupons to apply to specific offer(s)

Enables associating a coupon specifically with one or more offer
items, instead of coupons that could be applied to any offer
item. Deprecated, generic coupons will still work until they are
expired or updated to be specific.

blakeasher commented on app/models/coupon.rb:
```diff
@@ -129,18 +129,53 @@ def to_csv
     end
     raise CouponRanOut if coupon.redemptions_count >= coupon.how_many
     raise CouponExpired if coupon.expiration < Time.now.to_date
+    raise CouponNotApplicable if coupon.offers.present? && !coupon.offers.pluck(:code).include?(offer_code)
```
In case we'd like to change it -- this is the line that allows generic coupons to work until they run out / expire.

Sailias replied:
I like this implementation

blakeasher replied:
Awesome!

Sailias replied:
This should be a lookup on the offerable directly

Sailias commented on app/models/coupon.rb:
```diff
@@ -129,18 +129,53 @@ def to_csv
+      if !offer.valid?
+        self.errors.add(:offer, "code #{offer_code} is not valid")
+      end
+      offer
```
why return `offer` here?
Is it to ensure the return value isn't an empty array and the validation might fail if an empty array is returned because something truthy must be returned? If so, an empty array is truthy.

blakeasher replied:
Great question! It's so that this method returns an array of instantiated offers that are later saved [here](https://github.com/Sailias/couponsgem/blob/154b34fd7eee7ca630e245fbfb1dfd37328aeb24/app/controllers/concerns/couponing/crud_controller_methods.rb#L51). The `self.errors(..)` is what is actually checked for validity in the controller.

Without returning `offer`, this method would return an array of `nil`'s whether or not the offer is valid. But that does mean it also returns the same array of instantiated objects whether they are valid or not, which...doesn't make sense with what I'd expect either 🤔

blakeasher replied:
What do you think about changing it to something like this:
```ruby
offer_codes.map do |offer_code|
  offer = Offer.new(code: offer_code)
  if !offer.valid?
    self.errors.add(:offer, "code #{offer_code} is not valid")
    next
  else
    offer
  end
end
```

Sailias replied:
I think that this should be done via a before_validation or a accepts_nested_attributes_for.

Sailias replied:
because you already have `params[:offers]` it seems like you could do this with `accepts_nested_attributes_for` and let the offer validate itself on create and then display the errors in the form for the `offer`

blakeasher replied:
Yeah, this is a good call out. I wasn't able to make good use of `accepts_nested_attributes_for` for this context the way I was hoping -- but, I was still able to remove that `validate_new_offers` method! You’re right. We can let the offer validate itself.

blakeasher replied:
Came back with a refactor that does take advantage of `accepts_nested_attributes_for`. Let me know what you think!

Sailias commented on app/models/coupon.rb:
```diff
@@ -129,18 +129,53 @@ def to_csv
+    added_codes.each do |code|
+      self.offers << Offer.new(code: code)
+    end
+  end
```
You could use the _destroy form variable, but I see why you did it this way too.

blakeasher replied:
Yeah, I don't love it, but it worked...I'm open to doing it differently

blakeasher replied:
Looked into this again too. I refactored it some, but `:_destroy` didn't mesh well when `collection_check_boxes` is used and the view doesn't have offer ids.

blakeasher replied:
Also updated this to use `:_destroy`.

zingsurfer committed:
Refactor coupon offer validating

blakeasher commented on app/models/coupon.rb:
```diff
@@ -130,38 +131,23 @@ def can_edit?
+    Coupon.transaction do
+      self.update_offers(offer_codes)
+      self.update(coupon)
     end
```
Added this to ensure that offer deletions are rolled back if needed, for example when prior offer codes are removed but the coupon updates are not saved.

blakeasher replied:
Removed this -- it's taken care of for us now.

zingsurfer committed:
Remove unneeded variable assignment

zingsurfer committed:
Update offers table

zingsurfer committed:
Update coupon and offer behavior

zingsurfer committed:
Add offer specifics concern

Gathers offer-related stuff together here, so that it can be readily
included or overwritten as needed by apps that use this gem.

zingsurfer committed:
Update couponing gem controller actions

zingsurfer committed:
Refactor available offers

The prior method would not have worked when no offers have been stored.

zingsurfer committed:
Sweep nolonger needed methods

Sailias commented on app/controllers/concerns/couponing/crud_controller_methods.rb:
```diff
@@ -65,10 +68,13 @@ def edit
-        redirect_to coupon_redirect_path(after: @first_coupon)
-      else
+
+      @coupon.update coupon_params
```
why not just the regular `if @coupon.update coupon_params` why the extra `errors.messages.present?`

blakeasher replied:
You're right. There's no need for this change.

Sailias commented on app/controllers/concerns/couponing/crud_controller_methods.rb:
```diff
@@ -77,6 +83,11 @@ def update
+    end
+
+    def load_offers
+      @available_offers = @coupon.available_offers
```
This isn't necessary, a simple `@coupon.available_offers` in the view is fine

blakeasher replied:
Agreed, I'll update this.

Sailias commented on app/models/concerns/couponing/offer_specifics.rb:
```diff
@@ -0,0 +1,14 @@
+  end
+
+  def offer_list; end
+  def available_offers; end
```
This should return an empty array, not nil because you iterate over it.

blakeasher replied:
Good catch!

Sailias commented on app/models/coupon.rb:
```diff
@@ -103,9 +103,9 @@ def self.no_coupon(product_bag = {})
   # Return a hash with the new prices for each product, as well the grand total
   # and total savings
-  def self.apply(coupon_code, product_bag = {})
+  def self.apply(coupon_code, offer_code, product_bag = {})
```
You're still passing in the offer code, you should pass in the offerable.

blakeasher replied:
Oh woops, I missed this. I'll update it.

zingsurfer committed:
Revert unneeeded controller changes

zingsurfer committed:
Update applying a coupon to use offerable

zingsurfer committed:
Address more PR feedback
